src/navigation/reroute.js

```diff
@@ -93,90 +116,95 @@ export function reroute(pendingPromises = [], eventArguments) {
   function performAppChanges() {
     return Promise.resolve().then(() => {
       // https://github.com/single-spa/single-spa/issues/545
-      window.dispatchEvent(
-        new CustomEvent(
-          appsThatChanged.length === 0
-            ? "single-spa:before-no-app-change"
-            : "single-spa:before-app-change",
-          getCustomEventDetail(true)
-        )
+      fireSingleSpaEvent(
+        appsThatChanged.length === 0
+          ? "before-no-app-change"
+          : "before-app-change",
+        getCustomEventDetail(true)
       );
 
-      window.dispatchEvent(
-        new CustomEvent(
-          "single-spa:before-routing-event",
-          getCustomEventDetail(true, { cancelNavigation })
-        )
+      fireSingleSpaEvent(
+        "before-routing-event",
+        getCustomEventDetail(true, { cancelNavigation })
       );
 
-      if (navigationIsCanceled) {
-        window.dispatchEvent(
-          new CustomEvent(
-            "single-spa:before-mount-routing-event",
-            getCustomEventDetail(true)
-          )
-        );
-        finishUpAndReturn();
-        navigateToUrl(oldUrl);
-        return;
-      }
+      return Promise.all(cancelPromises).then((cancelValues) => {
+        const navigationIsCanceled = cancelValues.some((v) => v);
 
-      const unloadPromises = appsToUnload.map(toUnloadPromise);
+        if (navigationIsCanceled) {
+          // Change url back to old url, without triggering the normal single-spa reroute
+          originalReplaceState.call(
```

Comment:
Is it possible that originalReplaceState is `null` in any situation here?

Reply:
Hmm good catch. I think that if someone calls `navigateToUrl()` before calling `start()` or `patchHistoryApi()` that it would be undefined. I'll try to reproduce in a test, then fix to have it default to `window.history.replaceState` if `originalReplaceState` is not defined.

Reply:
After writing the test, it didn't throw an error. The reason is that `reroute()` doesn't allow you to cancel navigation until you've called `start()`, since it short circuits when single spa isn't started. And so this code will never execute until after patchHistoryApi (which is called by start()) is executed. See my upcoming commit in this PR that adds the test.